### PR TITLE
Setup Cart Page skeleton

### PR DESCRIPTION
**Pull request by rikxjeee**

### CODE REVIEW

File: src/Application/DefaultServiceContainer.php [hunk 1]
```diff
@@ -93,4 +97,20 @@ private function getUrlProvider(): UrlProvider
     {
         return new UrlProvider();
     }
+
+    private function getCartPageController(): Controller
+    {
+        return new CartPageController($this->getCart(), $this->getRenderer(),
```
Comment (tkotosz) [inline src/Application/DefaultServiceContainer.php]: all param should be in new line

File: src/Controller/CartPageController.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+use Supermarket\Model\Cart;
+use Supermarket\Model\Product;
+use Supermarket\Model\Request;
+use Supermarket\Model\Response;
+use Supermarket\Renderer\Renderer;
+use supermarket\Transformer\ProductToCartContentViewTransformer;
+
+class CartPageController implements Controller
+{
+    private const SUPPORTED_REQUEST = 'mycart';
```
Comment (tkotosz) [inline src/Controller/CartPageController.php]: cart should be more than enough :D

File: src/Controller/CartPageController.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        $this->renderer = $renderer;
+        $this->productToCartContentViewTransformer = $productToCartContentViewTransformer;
+    }
+
+    public function execute(Request $request): Response
+    {
+        /**
+         * one product is hardcoded until feature to add them manually is implemented
+         */
+        $this->cart->addToCart(new Product(1, 'cola', 1, 'drink'));
```
Comment (tkotosz) [inline src/Controller/CartPageController.php]: `addToCart` the word `Cart` is redundant... `add` or `addProduct` would be much better name for this method

File: src/Controller/CartPageController.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+        $cartContent = $this->cart->getCart();
```
Comment (tkotosz) [inline src/Controller/CartPageController.php]: `cart->getCart()` - :thinking:

File: src/Model/Cart.php [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+namespace Supermarket\Model;
+
+class Cart
+{
+    /**
+     * @var CartItem[]
+     */
+    private $cartItems = [];
+
+    public function getCart()
```
Comment (tkotosz) [inline src/Model/Cart.php]: getItems?

File: src/Model/Cart.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        return $this->cartItems;
+    }
+
+    public function addToCart(Product $product)
+    {
+        if(isset($this->cartItems[$product->getName()])) {
```
Comment (tkotosz) [inline src/Model/Cart.php]: use id instead

Comment (tkotosz) [reply src/Model/Cart.php]: also wrong fomatting: mising space after `if`

File: src/Model/Cart.php [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+            $this->cartItems[$product->getName()]->increaseQuantity();
+        }else{
```
Comment (tkotosz) [inline src/Model/Cart.php]: missing spaces around `else`

File: src/Model/CartItem.php [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * @return int
+     */
+    public function getQuantity(): int
+    {
+        return $this->quantity;
+    }
+
+    public function increaseQuantity()
```
Comment (tkotosz) [inline src/Model/CartItem.php]: missing return typehint

File: src/Model/CartItem.php [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    public function __construct(Product $product, int $quantity)
+    {
+        $this->product = $product;
+        $this->quantity = $quantity;
+    }
+
+    /**
+     * @return Product
+     */
+    public function getProduct(): Product
```
Comment (tkotosz) [inline src/Model/CartItem.php]: docblock can be removed

File: src/Model/CartItem.php [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        return $this->product;
+    }
+
+    /**
+     * @return int
+     */
+    public function getQuantity(): int
```
Comment (tkotosz) [inline src/Model/CartItem.php]: docblock can be removed

File: src/Model/View/CartContentView.php [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+class CartContentView
+{
+    /**
+     * @var CartItemView[]
+     */
+    private $items;
+
+    /**
+     * CartContentView constructor.
```
Comment (tkotosz) [inline src/Model/View/CartContentView.php]: this text doesn't add any value, can be removed

File: src/Model/View/CartContentView.php [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+     */
+    public function __construct(array $items)
+    {
+        $this->items = $items;
+    }
+
+    /**
+     * @return CartItemView[]
+     */
+    public function getCartContent(): array
```
Comment (tkotosz) [inline src/Model/View/CartContentView.php]: getItems?

File: src/Transformer/ProductToCartContentViewTransformer.php [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        $this->urlProvider = $urlProvider;
+    }
+
+    public function transform(Cart $cart): CartContentView
+    {
+        $cartItems = $cart->getCart();
+        $cartContentView = [];
+        foreach ($cartItems as $key => $cartItem) {
+            $cartContentView[] = new CartItemView($cartItem->getProduct()->getName(), $cartItem->getProduct()->getPrice(), $cartItem->getQuantity());
```
Comment (tkotosz) [inline src/Transformer/ProductToCartContentViewTransformer.php]: too long line

File: src/Transformer/ProductToCartContentViewTransformer.php [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+    public function __construct(UrlProvider $urlProvider)
+    {
+        $this->urlProvider = $urlProvider;
+    }
+
+    public function transform(Cart $cart): CartContentView
+    {
+        $cartItems = $cart->getCart();
+        $cartContentView = [];
```
Comment (tkotosz) [inline src/Transformer/ProductToCartContentViewTransformer.php]: cartContentViewData

File: src/template/cart/empty_cart.html [hunk 15]
```diff
@@ -0,0 +1,2 @@
+<a href="index.php">Back</a>
```
Comment (tkotosz) [inline src/template/cart/empty_cart.html]: `index.php` should not be hard-coded

File: src/template/index.html [hunk 16]
```diff
@@ -5,6 +5,10 @@
     <title>Supermarket</title>
 </head>
 <body>
+<a href="index.php?page=mycart">My Cart</a>
```
Comment (tkotosz) [inline src/template/index.html]: url should no be hard-coded here

File: src/template/index.html [hunk 17]
```diff
@@ -5,6 +5,10 @@
...
+<a href="index.php?page=products">Products</a>
```
Comment (tkotosz) [inline src/template/index.html]: url should no be hard-coded here

File: src/Application/DefaultServiceContainer.php [hunk 18]
```diff
@@ -93,4 +97,20 @@ private function getUrlProvider(): UrlProvider
...
+            $this->getProductToCartContentViewTransformer());
+    }
+
+    private function getCart(): Cart
+    {
+        return new Cart();
+    }
+
+    private function getProductToCartContentViewTransformer()
```
Comment (tkotosz) [inline src/Application/DefaultServiceContainer.php]: missing return typehint

File: src/Application.php [hunk 19]
```diff
@@ -21,6 +22,10 @@ public function __construct(ServiceContainer $serviceContainer)
 
     public function run(): void
     {
+        if (!isset($_SESSION['cart'])) {
```
Comment (tkotosz) [inline src/Application.php]: $_SESSION should be wrapped in a object like we did with the Request so you can call get and start on it

File: src/Controller/CartPageController.php [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+        if(empty($cartContent)) {
```
Comment (tkotosz) [inline src/Controller/CartPageController.php]: this is rendering logic, can be moved to the renderer (just need to pass over the empty cart template along with the other ones

File: index.php [hunk 21]
```diff
@@ -9,4 +9,6 @@
     throw new Exception('configuration missing');
 }
 
+session_start();
```
Comment (tkotosz) [inline index.php]: :see_no_evil:

File: src/Application/DefaultServiceContainer.php [hunk 22]
```diff
@@ -93,4 +97,20 @@ private function getUrlProvider(): UrlProvider
...
+
+    private function getCartPageController(): Controller
+    {
+        return new CartPageController($this->getCart(), $this->getRenderer(),
+            $this->getProductToCartContentViewTransformer());
+    }
+
+    private function getCart(): Cart
+    {
+        return new Cart();
```
Comment (tkotosz) [inline src/Application/DefaultServiceContainer.php]: Cart is a Model therefore should not be requested or injected as a service...
You are missing to have a provider which can return the _current cart_ which should work like this:
1. get the current cart id from session
2. get the cart based on the cart id <-- for this you need a cart repository like the product repository

File: src/Application.php [hunk 23]
```diff
@@ -14,23 +15,42 @@ class Application
...
                 ->getRouter()
                 ->match($request)
-                ->execute($request);
+                ->execute($request, $this->session);
         } catch (Exception $e) {
             $response = new Response($e->getMessage(), Response::STATUS_SERVER_ERROR);
         }
         http_response_code($response->getStatusCode());
         echo $response->getContent();
+        Session::destroy();
```
Comment (tkotosz) [inline src/Application.php]: :D

File: src/Application.php [hunk 24]
```diff
@@ -14,23 +15,42 @@ class Application
...
+
     public function __construct(ServiceContainer $serviceContainer)
     {
         $this->serviceContainer = $serviceContainer;
     }
 
     public function run(): void
     {
+        if (!isset($_SESSION)) {
+            Session::start();
```
Comment (tkotosz) [inline src/Application.php]: convert te session to a service

File: src/Application/DefaultServiceContainer.php [hunk 25]
```diff
@@ -84,13 +78,51 @@ private function getProductsToProductListViewTransformer(): ProductsToProductLis
...
+    {
+        return new ProductToCartContentViewTransformer($this->getUrlProvider());
+    }
+
+    public function getSessionManager()
+    {
+        return new SessionManager();
+    }
+
+    public function getDataBaseBasedCartRepository(): DatabaseBasedCartRepository
```
Comment (tkotosz) [inline src/Application/DefaultServiceContainer.php]: should be private

File: src/Application/DefaultServiceContainer.php [hunk 26]
```diff
@@ -84,13 +78,51 @@ private function getProductsToProductListViewTransformer(): ProductsToProductLis
...
+            $this->getDataBaseBasedCartRepository()
+        );
+    }
+
+    private function getProductToCartContentViewTransformer(): ProductToCartContentViewTransformer
+    {
+        return new ProductToCartContentViewTransformer($this->getUrlProvider());
+    }
+
+    public function getSessionManager()
```
Comment (tkotosz) [inline src/Application/DefaultServiceContainer.php]: public move up

File: src/Application/DefaultServiceContainer.php [hunk 27]
```diff
@@ -84,13 +78,51 @@ private function getProductsToProductListViewTransformer(): ProductsToProductLis
         return new ProductsToProductListViewTransformer($this->getUrlProvider());
     }
 
+    private function getUrlProvider(): UrlProvider
+    {
+        return new UrlProvider();
+    }
+
+    public function getProductDetailsController(): Controller
```
Comment (tkotosz) [inline src/Application/DefaultServiceContainer.php]: public move up

File: src/Application/SessionManager.php [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        session_start();
+    }
+
+    public static function destroy(): void
+    {
+        session_destroy();
+    }
+
+    public function getField(string $param)
```
Comment (tkotosz) [inline src/Application/SessionManager.php]: missing return typehint `?string`

File: src/Application/SessionManager.php [hunk 29]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        return $_SESSION[$param];
```
Comment (tkotosz) [inline src/Application/SessionManager.php]: might throw undefined index error... use ?? '' or ?? null

File: src/Controller/CartPageController.php [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+    ) {
+        $this->renderer = $renderer;
+        $this->productToCartContentViewTransformer = $productToCartContentViewTransformer;
+        $this->sessionManager = $sessionManager;
+        $this->databaseBasedCartRepository = $databaseBasedCartRepository;
+    }
+
+    public function execute(Request $request): Response
+    {
+        $this->sessionManager::start();
```
Comment (tkotosz) [inline src/Controller/CartPageController.php]: ???

File: src/Application/SessionManager.php [hunk 31]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace Supermarket\Application;
+
+class SessionManager
+{
+    public static function start(): void
```
Comment (tkotosz) [inline src/Application/SessionManager.php]: should not be static

File: src/Application/SessionManager.php [hunk 32]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        session_start();
+    }
+
+    public static function destroy(): void
```
Comment (tkotosz) [inline src/Application/SessionManager.php]: should not be static

File: src/Controller/CartPageController.php [hunk 33]
```diff
@@ -0,0 +1,85 @@
...
+        $this->sessionManager->addField('user_id', 2);
```
Comment (tkotosz) [inline src/Controller/CartPageController.php]: it doesn't add value, it sets it so lets call it set... also it is setting a value not a field setValue getValue generall sounds better

Comment (tkotosz) [inline src/Controller/CartPageController.php]: also why would you set anything to the session here?

File: src/Controller/CartPageController.php [hunk 34]
```diff
@@ -0,0 +1,85 @@
...
+        $cartId = $this->sessionManager->getField('user_id');
+
+        $this->cart = $this->databaseBasedCartRepository->getCartById($cartId);
+        $this->sessionManager->addField('cart_id', $this->cart->getCartId());
```
Comment (tkotosz) [inline src/Controller/CartPageController.php]: again why is this needed?

File: src/Controller/CartPageController.php [hunk 35]
```diff
@@ -0,0 +1,85 @@
...
+        $this->databaseBasedCartRepository = $databaseBasedCartRepository;
+    }
+
+    public function execute(Request $request): Response
+    {
+        $this->sessionManager::start();
+        $this->sessionManager->addField('user_id', 2);
+        $cartId = $this->sessionManager->getField('user_id');
+
+        $this->cart = $this->databaseBasedCartRepository->getCartById($cartId);
```
Comment (tkotosz) [inline src/Controller/CartPageController.php]: need to check if cartId empty before passing it to the repo

File: src/Controller/CartPageController.php [hunk 36]
```diff
@@ -0,0 +1,85 @@
...
+
+        $cartContent = $this->cart->getItems();
+        if (empty($cartContent)) {
+            $content = $this->renderer->renderEmptyCart('/cart/empty_cart.html');
+
+            return new Response($content);
+        }
+        $cartItemsView = $this->productToCartContentViewTransformer->transform($this->cart);
+        $content = $this->renderer->renderCart(
+            $cartItemsView,
```
Comment (tkotosz) [inline src/Controller/CartPageController.php]: pass over a cartView instead, then the above empty cart page rendering logic can be moved to the renderer

File: src/Model/Cart.php [hunk 37]
```diff
@@ -0,0 +1,38 @@
...
+    public function addProduct(Product $product)
+    {
+        if (isset($this->cartItems[$product->getId()])) {
+            $this->cartItems[$product->getId()]->increaseQuantity();
+        } else {
+            $this->cartItems[$product->getId()] = new CartItem($product);
+        }
+    }
+
+    public function getCartId(): int
```
Comment (tkotosz) [inline src/Model/Cart.php]: move this up under the constuctor since it is a property getter

File: src/Model/Config/ApplicationConfig.php [hunk 38]
```diff
@@ -21,6 +21,12 @@ class ApplicationConfig
      */
     private $templateConfig;
 
+    private function __construct(DatabaseConfig $databaseCredentials, TemplateConfig $templateConfig)
```
Comment (tkotosz) [inline src/Model/Config/ApplicationConfig.php]: move this down since it is private

File: src/Model/Config/ApplicationConfig/TemplateConfig.php [hunk 39]
```diff
@@ -9,6 +9,11 @@ class TemplateConfig
      */
     private $basePath;
 
+    private function __construct(string $basePath)
```
Comment (tkotosz) [inline src/Model/Config/ApplicationConfig/TemplateConfig.php]: move this down, snice it is private

File: src/Model/View/CartItemView/CartItemView.php [hunk 40]
```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace Supermarket\Model\View\CartItemView;
+
+class CartItemView
```
Comment (tkotosz) [inline src/Model/View/CartItemView/CartItemView.php]: namespace and class name is the same... redundant

File: src/Renderer/HTMLrenderer.php [hunk 41]
```diff
@@ -29,24 +30,52 @@ public function renderProductListTable(
         return $this->renderTemplate(['content' => $list], $this->loadTemplate('index.html'));
     }
 
-    public function renderProductDetails(ProductDetailsView $product, string $productDetailsTemplate): string
+    private function renderTemplate(array $data, string $content): string
```
Comment (tkotosz) [inline src/Renderer/HTMLrenderer.php]: private -> move it down

File: src/Repository/DatabaseBasedCartRepository.php [hunk 42]
```diff
@@ -0,0 +1,39 @@
...
+
+    public function getCartById(?int $id): Cart
+    {
+        if ($id === null) {
+            return new Cart(uniqid());
+        }
+
+        $fetchCart = $this->mySqlConnection->prepare(
+            'SELECT * FROM products_in_carts
+        LEFT JOIN products
```
Comment (tkotosz) [inline src/Repository/DatabaseBasedCartRepository.php]: holy cow ! :D

File: src/Controller/CartPageController.php (hunk 30 above)
Comment (rikxjeee) [reply src/Controller/CartPageController.php]: Should i start the session as soon as the sessionmanager created by the service container?

File: src/Controller/CartPageController.php (hunk 33 above)
Comment (rikxjeee) [reply src/Controller/CartPageController.php]: For emulating multiple users to manually test some scenarios - basically switching carts on the fly by changing the fake users id

File: src/Controller/CartPageController.php [hunk 43]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        $this->sessionManager->start();
+        $this->sessionManager->setValue('cart_id', 1);
+        $cartId = $this->sessionManager->getValue('cart_id');
+
+        /**
+         * carts are statically stored in db until feature to add them manually is implemented
+         */
+        $this->cart = $this->databaseBasedCartRepository->getCartById($cartId);
+        $this->sessionManager->setValue('cart_id', $this->cart->getCartId());
```
Comment (tkotosz) [inline src/Controller/CartPageController.php]: ??

File: src/Controller/CartPageController.php (hunk 30 above)
Comment (tkotosz) [reply src/Controller/CartPageController.php]: I am ok with starting the session here, just change the method to be an instance method rather than a static one

File: src/Provider/UrlProvider.php [hunk 44]
```diff
@@ -9,8 +9,13 @@ public function getProductUrl(int $id): string
         return sprintf('index.php?page=details&id=%d', $id);
     }
 
-    public function getProductListUrl()
+    public static function getProductListUrl(): string
```
Comment (tkotosz) [inline src/Provider/UrlProvider.php]: why static?

File: src/Provider/UrlProvider.php [hunk 45]
```diff
@@ -9,8 +9,13 @@ public function getProductUrl(int $id): string
...
     {
         return 'index.php?page=products';
     }
+
+    public static function getCartUrl(): string
```
Comment (tkotosz) [inline src/Provider/UrlProvider.php]: same here
